File: packages/typescript/src/client/resources/SeriesResource.ts
import type { components } from "../../generated/api";
import type { Page, PaginationOptions } from "../pagination/Page";
import { BaseResource } from "./BaseResource";

// Type aliases
type Series = components["schemas"]["Series"];
type SeriesType = components["schemas"]["SeriesType"];
type TcgType = components["schemas"]["TcgType"];

// Response types
type SeriesListResponse = components["schemas"]["SeriesListResponse"];
type SeriesSingleResponse = components["schemas"]["SeriesSingleResponse"];

/** Options for listing series */
export interface ListSeriesOptions extends PaginationOptions {
  /** Filter by TCG type (Pokemon, Magic, YuGiOh) */
  tcgType?: TcgType;
  /** Filter by series type (Standard, Special, Promo) */
  seriesType?: SeriesType;
}

/** Series API - browse card series */
export class SeriesResource extends BaseResource {
  /**
   * Get all series with optional filtering and pagination.
   *
   * @example
   * ```ts
   * // Get first page of series
   * const page = await client.series.list();
   *
   * // Filter by type
   * const standardSeries = await client.series.list({
   *   seriesType: 'Standard'
   * });
   *
   * // Iterate through all series
   * for await (const s of client.series.list()) {
   *   console.log(s.name);
   * }
   * ```
   */
  async list(options: ListSeriesOptions = {}): Promise<Page<Series>> {
    const query = this.buildQuery({
      page: options.page,
      pageSize: options.pageSize,
      tcgType: options.tcgType,
      seriesType: options.seriesType,
    });

    const response = await this.http.get<SeriesListResponse>("/Series", query);

    return this.createPagedResponse(response, (paginationOpts) =>
      this.list({ ...options, ...paginationOpts })
    );
  }

  /**
   * Get a single series by ID.
   *
   * @param id - The series GUID
   */
  async get(id: string): Promise<Series> {
    const response = await this.http.get<SeriesSingleResponse>(`/Series/${id}`);

    if (!response.data) {
      throw new Error("Series not found");
    }

    return response.data;
  }

  /**
   * Get a series by its URL-friendly slug.
   *
   * @param slug - The URL-friendly slug (e.g., "scarlet-violet")
   */
  async getBySlug(slug: string): Promise<Series> {
    const response = await this.http.get<SeriesSingleResponse>(
      `/Series/${slug}`
    );

    if (!response.data) {
      throw new Error("Series not found");
    }

    return response.data;
  }

  /**
   * Convenience method to iterate through all series.
   *
   * @example
   * ```ts
   * for await (const s of client.series.listAll()) {
   *   console.log(s.name);
   * }
   * ```
   */
  async *listAll(
    options: Omit<ListSeriesOptions, "page"> = {}
  ): AsyncGenerator<Series> {
    const page = await this.list({ ...options, page: 1 });
    yield* page;
  }
}

File: packages/csharp/src/PokeForge.SDK/Client/Resources/SeriesResource.cs
namespace PokeForge.SDK.Client.Resources;

using System.Runtime.CompilerServices;
using PokeForge.SDK.Client.Http;
using PokeForge.SDK.Client.Pagination;
using PokeForge.SDK.Exceptions;
using PokeForge.SDK.Generated;

/// <summary>
/// Options for listing series.
/// </summary>
public sealed class ListSeriesOptions : PaginationOptions
{
    /// <summary>
    /// Sort field.
    /// </summary>
    public string? SortBy { get; set; }

    /// <summary>
    /// Sort order (asc or desc).
    /// </summary>
    public string? SortOrder { get; set; }

    /// <summary>
    /// Search by series name.
    /// </summary>
    public string? Search { get; set; }

    internal ListSeriesOptions Clone() => new()
    {
        Page = Page,
        PageSize = PageSize,
        SortBy = SortBy,
        SortOrder = SortOrder,
        Search = Search
    };
}

/// <summary>
/// Series API - browse card series.
/// </summary>
public sealed class SeriesResource : BaseResource
{
    internal SeriesResource(HttpClientWrapper http) : base(http)
    {
    }

    /// <summary>
    /// Get all series with optional filtering and pagination.
    /// </summary>
    public async Task<Page<SeriesList>> ListAsync(
        ListSeriesOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        options ??= new ListSeriesOptions();

        var query = BuildQuery(
            ("page", options.Page),
            ("pageSize", options.PageSize),
            ("sortBy", options.SortBy),
            ("sortOrder", options.SortOrder),
            ("search", options.Search)
        );

        var response = await Http.GetAsync<SeriesListResponse>("/Series", query, cancellationToken)
            .ConfigureAwait(false);

        return CreatePagedResponse(
            response!,
            r => r.Data,
            r => r.Pagination,
            async (page, pageSize, ct) =>
            {
                var newOptions = options.Clone();
                newOptions.Page = page;
                newOptions.PageSize = pageSize;
                return await ListAsync(newOptions, ct).ConfigureAwait(false);
            });
    }

    /// <summary>
    /// Get a single series by ID with full details including sets.
    /// </summary>
    public async Task<SeriesDetail> GetAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var response = await Http.GetAsync<SeriesSingleResponse>($"/Series/{id}", cancellationToken: cancellationToken)
            .ConfigureAwait(false);

        return response?.Data ?? throw new NotFoundException("Series not found");
    }

    /// <summary>
    /// Iterate through all series using IAsyncEnumerable.
    /// </summary>
    public async IAsyncEnumerable<SeriesList> ListAllAsync(
        ListSeriesOptions? options = null,
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        options ??= new ListSeriesOptions();
        options.Page = 1;

        var page = await ListAsync(options, cancellationToken).ConfigureAwait(false);

        await foreach (var series in page.WithCancellation(cancellationToken))
        {
            yield return series;
        }
    }
}
